develop/detail/macro_scope.hpp

```diff
@@ -41,9 +41,21 @@
     #define JSON_TRY try
     #define JSON_CATCH(exception) catch(exception)
 #else
-    #define JSON_THROW(exception) std::abort()
-    #define JSON_TRY if(true)
-    #define JSON_CATCH(exception) if(false)
+    #if !defined(JSON_THROW)
+        #define JSON_THROW(exception) std::abort()
```

Comment:
I just realized the code can be simplified:

```cpp
#if !defined(JSON_THROW)
    #define JSON_THROW(exception) std::abort()
    #define JSON_THROW_DEFAULT
#endif
```

and later

```cpp
#if defined(JSON_THROW_DEFAULT)
    #undef JSON_THROW
    #undef JSON_THROW_DEFAULT
#endif
```

What do you think?

Reply:
Yes, this is better, less noise

Reply:
But then it requires the _DEFAULT defines for exceptions-on version as well, ends up adding a bit of clutter

Reply:
Right... But why not allowing to define `JSON_THROW` etc. while using exceptions?

Reply:
The current version in repo does #undef for both exceptions-on and exceptions-off cases. If i rewrite with JSON_THROW_DEFAULT then to keep the same behaviour it would become ..

```
#if (defined(__cpp_exceptions) ...
    #define JSON_THROW(exception) throw exception
    #define JSON_THROW_DEFAULT //  1 extra line here
   ...
#else
   #if !defined(JSON_THROW)
    #define JSON_THROW(exception) std::abort()
    #define JSON_THROW_DEFAULT
  #endif
  ...
#endif
```
and later
```
#if defined(JSON_THROW_DEFAULT)
    #undef JSON_THROW
    #undef JSON_THROW_DEFAULT
#endif
```

If that is cleaner, I'll rewrite

Reply:
Another idea:

```cpp
// allow to disable exceptions
#if (defined(__cpp_exceptions) || defined(__EXCEPTIONS) || defined(_CPPUNWIND)) && !defined(JSON_NOEXCEPTION)
    #define JSON_THROW(exception) throw exception
    #define JSON_TRY try
    #define JSON_CATCH(exception) catch(exception)
#else
    #define JSON_THROW(exception) std::abort()
    #define JSON_TRY if(true)
    #define JSON_CATCH(exception) if(false)
#endif

// override exception macros
#if define(JSON_THROW_USER)
    #define JSON_THROW JSON_THROW_USER
#endif
#if define(JSON_TRY_USER)
    #define JSON_TRY JSON_TRY_USER
#endif
#if define(JSON_CATCH_USER)
    #define JSON_CATCH JSON_CATCH_USER
#endif
```

That is, the user can define `JSON_THROW_USER` to override the behavior of `JSON_THROW`. No need to differentiate the exception/no exception case, and no need to remember whether we need to undef.

What do you think?

Reply:
This is probably the most hygienic , although for maximum warning compile it should be 
```
#if define(JSON_THROW_USER)
    #undef JSON_THROW
    #define JSON_THROW JSON_THROW_USER
#endif
```
to avoid macro redefinition warnings

Reply:
Right.